File: code/07-hash-tables/java/StringHash.java
/** StringHash: Horner's rule 다항식 문자열 해시. hash = hash*B + code(c)를
 * 왼쪽에서 오른쪽으로 누적한다(lecture-notes/lecture07/sections/04_string_hashing.tex). */
public final class StringHash {
    // snippet:string-hash:start
    public static int asciiSum(String s) {
        int sum = 0;
        for (int i = 0; i < s.length(); i++) sum += s.charAt(i);
        return sum;
    }

    public static long stringHash(String s, int base) {
        long hash = 0;
        for (int i = 0; i < s.length(); i++) hash = hash * base + s.charAt(i);
        return hash;
    }
    // snippet:string-hash:end

    public static void main(String[] args) {
        System.out.println("ascii_sum(abcd): " + asciiSum("abcd"));
        System.out.println("ascii_sum(dbac): " + asciiSum("dbac"));
        System.out.println("string_hash(abcd): " + stringHash("abcd", 131));
        System.out.println("string_hash(dbac): " + stringHash("dbac", 131));
        System.out.println("string_hash(Apple): " + stringHash("Apple", 131));
        System.out.println("string_hash(Apply): " + stringHash("Apply", 131));
    }
}

File: code/07-hash-tables/c/string_hash.c
#include <stdio.h>
#include <string.h>

/* StringHash: Horner's rule 다항식 문자열 해시. hash = hash*B + code(c)를
 * 왼쪽에서 오른쪽으로 누적한다(lecture-notes/lecture07/sections/04_string_hashing.tex). */

// snippet:string-hash:start
static long ascii_sum(const char *s) {
    long sum = 0;
    for (size_t i = 0; s[i] != '\0'; i++) sum += (unsigned char)s[i];
    return sum;
}

static long long string_hash(const char *s, long long base) {
    long long hash = 0;
    for (size_t i = 0; s[i] != '\0'; i++) hash = hash * base + (unsigned char)s[i];
    return hash;
}
// snippet:string-hash:end

int main(void) {
    printf("ascii_sum(abcd): %ld\n", ascii_sum("abcd"));
    printf("ascii_sum(dbac): %ld\n", ascii_sum("dbac"));
    printf("string_hash(abcd): %lld\n", string_hash("abcd", 131));
    printf("string_hash(dbac): %lld\n", string_hash("dbac", 131));
    printf("string_hash(Apple): %lld\n", string_hash("Apple", 131));
    printf("string_hash(Apply): %lld\n", string_hash("Apply", 131));
    return 0;
}
